File: config/config-api/src/com/thoughtworks/go/domain/scm/SCMs.java
/*************************GO-LICENSE-START*********************************
 * Copyright 2014 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *************************GO-LICENSE-END***********************************/

package com.thoughtworks.go.domain.scm;

import com.thoughtworks.go.config.ConfigCollection;
import com.thoughtworks.go.config.ConfigTag;
import com.thoughtworks.go.config.Validatable;
import com.thoughtworks.go.config.ValidationContext;
import com.thoughtworks.go.domain.BaseCollection;
import com.thoughtworks.go.domain.ConfigErrors;
import com.thoughtworks.go.util.ListUtil;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

import static com.thoughtworks.go.util.ListUtil.join;

@ConfigTag("scms")
@ConfigCollection(value = SCM.class)
public class SCMs extends BaseCollection<SCM> implements Validatable {

    public SCMs() {
    }

    public SCMs(SCM... scms) {
        Collections.addAll(this, scms);
    }

    public SCM find(final String scmId) {
        return ListUtil.find(this, new ListUtil.Condition() {
            @Override
            public <T> boolean isMet(T scm) {
                return ((SCM) scm).getId().equals(scmId);
            }
        });
    }

    @Override
    public void validate(ValidationContext validationContext) {
        validateNameUniqueness();
        validateFingerprintUniqueness();
    }

    @Override
    public ConfigErrors errors() {
        return new ConfigErrors();
    }

    @Override
    public void addError(String fieldName, String message) {
        throw new RuntimeException("Not Implemented");
    }

    public void removeSCM(String id) {
        SCM scmToBeDeleted = this.find(id);
        if (scmToBeDeleted == null) {
            throw new RuntimeException(String.format("Could not find SCM with id '%s'", id));
        }
        this.remove(scmToBeDeleted);
    }

    private void validateNameUniqueness() {
        HashMap<String, SCMs> map = new HashMap<>();

        for (SCM scm : this) {
            String name = scm.getName().toLowerCase();
            if (!map.containsKey(name)) {
                map.put(name, new SCMs());
            }
            map.get(name).add(scm);
        }

        for (String name : map.keySet()) {
            SCMs scmsWithSameName = map.get(name);
            if (scmsWithSameName.size() > 1) {
                for (SCM scm : scmsWithSameName) {
                    scm.addError(SCM.NAME, String.format("Cannot save SCM, found multiple SCMs called '%s'. SCM names are case-insensitive and must be unique.", scm.getName()));
                }
            }
        }
    }

    private void validateFingerprintUniqueness() {
        HashMap<String, SCMs> map = new HashMap<>();

        for (SCM scm : this) {
            String fingerprint = scm.getFingerprint();
            if (!map.containsKey(fingerprint)) {
                map.put(fingerprint, new SCMs());
            }
            map.get(fingerprint).add(scm);
        }

        for (String fingerprint : map.keySet()) {
            SCMs scmsWithSameFingerprint = map.get(fingerprint);
            if (scmsWithSameFingerprint.size() > 1) {
                List<String> scmNames = new ArrayList<>();
                for (SCM scm : scmsWithSameFingerprint) {
                    scmNames.add(scm.getName());
                }

                for (SCM scm : scmsWithSameFingerprint) {
                    scm.addError(SCM.SCM_ID, String.format("Cannot save SCM, found duplicate SCMs. %s", join(scmNames)));
                }
            }
        }
    }
}

File: server/webapp/WEB-INF/rails.new/app/assets/new_javascripts/models/pipeline_configs/scms.js
/*
 * Copyright 2016 ThoughtWorks, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */


define(['mithril', 'lodash', 'string-plus', 'models/model_mixins', 'helpers/mrequest', 'models/errors', 'models/pipeline_configs/encrypted_value',
  'models/validatable_mixin', 'js-routes'],
  function (m, _, s, Mixins, mrequest, Errors, EncryptedValue, Validatable, Routes) {
    var SCMs = m.prop([]);
    SCMs.scmIdToEtag = {};

    function plainOrCipherValue(data) {
      if (data.encrypted_value) {
        return new EncryptedValue({cipherText: s.defaultToIfBlank(data.encrypted_value, '')});
      } else {
        return new EncryptedValue({clearText: s.defaultToIfBlank(data.value, '')});
      }
    }

    SCMs.SCM = function (data) {
      Validatable.call(this, data);

      this.init = function (data) {
        this.id             = m.prop(s.defaultToIfBlank(data.id));
        this.name           = m.prop(s.defaultToIfBlank(data.name, ''));
        this.autoUpdate     = m.prop(s.defaultToIfBlank(data.auto_update));
        this.pluginMetadata = m.prop(new SCMs.SCM.PluginMetadata(data.plugin_metadata || {}));
        this.configuration  = s.collectionToJSON(m.prop(SCMs.SCM.Configurations.fromJSON(data.configuration || {})));
        this.errors         = m.prop(new Errors(data.errors));
      };

      this.init(data);

      this.reInitialize = function (data) {
        this.init(data);
      };

      this.clone = function () {
        return new SCMs.SCM(JSON.parse(JSON.stringify(this)));
      };

      this.toJSON = function() {
        /* eslint-disable camelcase */
        return {
          id:              this.id(),
          name:            this.name(),
          auto_update:     this.autoUpdate(),
          plugin_metadata: this.pluginMetadata().toJSON(),
          configuration:   this.configuration
        };
        /* eslint-enable camelcase */
      };

      this.update = function() {
        var self = this;

        var config =  function (xhr) {
          xhr.setRequestHeader("Content-Type", "application/json");
          xhr.setRequestHeader("Accept", "application/vnd.go.cd.v1+json");
          xhr.setRequestHeader("If-Match", SCMs.scmIdToEtag[self.id()]);
        };

        var extract = function(xhr) {
          if(xhr.status === 200) {
            SCMs.scmIdToEtag[self.id()] = xhr.getResponseHeader('ETag');
          }
          return xhr.responseText;
        };

        return m.request({
          method:     'PATCH',
          url:        Routes.apiv1AdminScmPath({material_name: this.name()}), //eslint-disable-line camelcase
          background: false,
          config:     config,
          extract:    extract,
          data:       this,
          type:       SCMs.SCM
        });
      };

      this.create = function() {
        var extract = function(xhr) {
          if(xhr.status === 200) {
            SCMs.scmIdToEtag[JSON.parse(xhr.responseText).id] = xhr.getResponseHeader('ETag');
          }
          return xhr.responseText;
        };

        return m.request({
          method:     'POST',
          url:        Routes.apiv1AdminScmsPath(),
          background: false,
          config:     mrequest.xhrConfig.v1,
          extract:    extract,
          data:       this,
          type:       SCMs.SCM
        });
      };
    };

    SCMs.SCM.PluginMetadata = function (data) {
      this.id      = m.prop(s.defaultToIfBlank(data.id, ''));
      this.version = m.prop(s.defaultToIfBlank(data.version, ''));

      this.toJSON = function () {
        return {
          id:      this.id(),
          version: this.version()
        };
      };
    };

    SCMs.SCM.Configurations = function (data) {
      Mixins.HasMany.call(this, {
        factory:    SCMs.SCM.Configurations.create,
        as:         'Configuration',
        collection: data
      });

      function configForKey(key) {
        return this.findConfiguration(function (config) {
          return _.isEqual(config.key(), key);
        });
      }

      this.valueFor = function (key) {
        var config = configForKey.call(this, key);

        if (config) {
          return config.value();
        }
      };

      this.setConfiguration = function (key, value) {
        var existingConfig = configForKey.call(this, key);

        if (!existingConfig) {
          return this.createConfiguration({key: key, value: value});
        }

        if (existingConfig.isSecureValue()) {
          if (!_.isEqual(value, existingConfig.value())) {
            existingConfig.becomeUnSecureValue();
            existingConfig.value(value);
          }
        } else {
          existingConfig.value(value);
        }
      };
    };

    SCMs.SCM.Configurations.create = function (data) {
      return new SCMs.SCM.Configurations.Configuration(data);
    };

    SCMs.SCM.Configurations.fromJSON = function (data) {
      var configurations = _.map(data, function (d) {
        return new SCMs.SCM.Configurations.Configuration(d);
      });

      return new SCMs.SCM.Configurations(configurations);
    };

    SCMs.SCM.Configurations.Configuration = function (data) {
      this.parent = Mixins.GetterSetter();
      this.key    = m.prop(s.defaultToIfBlank(data.key, ''));
      var _value  = m.prop(plainOrCipherValue(data));

      Mixins.HasEncryptedAttribute.call(this, {attribute: _value, name: 'value'});

      this.toJSON = function () {
        var valueHash = this.isPlainValue() ? {value: this.value()} : {encrypted_value: this.value()};  //eslint-disable-line camelcase

        return _.merge({key: this.key()}, valueHash);
      };
    };

    SCMs.init = function () {
      return m.request({
        method:        'GET',
        url:           Routes.apiv1AdminScmsPath(),
        config:        mrequest.xhrConfig.v1,
        type:          SCMs.SCM,
        unwrapSuccess: function (response) {
          return response._embedded.scms;
        }
      }).then(SCMs);
    };

    SCMs.filterByPluginId = function (pluginId) {
      return _.filter(SCMs(), function (scm) {
        return _.isEqual(scm.pluginMetadata().id(), pluginId);
      });
    };

    SCMs.findById = function (id) {
      var scm = _.find(SCMs(), function (scm) {
        return _.isEqual(scm.id(), id);
      });

      if(_.isNil(scm)) {
        return null;
      }

      var extract = function(xhr) {
        SCMs.scmIdToEtag[scm.id()] = xhr.getResponseHeader('ETag');
        return xhr.responseText;
      };

      return m.request({
        method:     'GET',
        url:        Routes.apiv1AdminScmPath({material_name: scm.name()}),  //eslint-disable-line camelcase
        background: false,
        config:     mrequest.xhrConfig.v1,
        extract:    extract,
        type:       SCMs.SCM
      });
    };

    SCMs.vm = function () {
      this.saveState = m.prop('');
      var errors    = [];

      this.startUpdating = function () {
        errors = [];
        this.saveState('in-progress disabled');
      };

      this.saveFailed = function (data) {
        errors.push(data.message);

        if(data.data) {
          if(data.data.configuration) {
            errors = _.concat(errors, _.flattenDeep(_.map(data.data.configuration, function(conf) {return _.values(conf.errors);})));
          }
        }

        this.saveState('alert');
      };

      this.saveSuccess = function () {
        this.saveState('success');
      };

      this.clearErrors = function () {
        errors = [];
      };

      this.reset = function () {
        errors = [];
        this.saveState('');
      };

      this.errors = function () {
        return errors;
      };

      this.hasErrors = function () {
        return !_.isEmpty(errors);
      };

      this.markClientSideErrors = function () {
        errors.push('There are errors on the page, fix them and save');
      };
    };

    return SCMs;
  });